core/src/data_sources/data_source.rs

```diff
@@ -404,21 +436,36 @@ impl DataSource {
                 &remove_tags,
             )
             .await?;
+        self.upsert_field_to_document_payload(qdrant_client, document_id, "tags", new_tags.clone())
+            .await?;
+        Ok(new_tags)
+    }
+
+    async fn upsert_field_to_document_payload(
```

Comment:
I think update is stlightly better naming than upsert here. We're really updating not upserting as the field already exist always

Reply:
Not sure I get you here; the function is effectively doing an upsert -- and the field does not exist yet in the payload, so at least during the field population migration it will be created by the function. 

In general, I would name functions according to the semantics of what they do, rather than how callers use it (of course the best is when we are sure that the two match). Maybe I'm missing something here?

Reply:
The document already exists you just update one of its fields

Reply:
Oh ok sorry for the misunderstanding. It's an upsert in the payload since the field is created or updated: I could call it create_or_update_field_in_document_payload => upsert_field_in_document_payload 

but it is an _update_ of the document's payload if we consider the payload a field

sql abstractions match less with nosql db (which qdrant got away with by calling it differently, "set_payload", not very clearly though)

I can call it `update_document_payload` which sounds clean

Reply:
`upsert` suggests that something is possibly going to be created. Nothing is created here, the document exists. right?

Reply:
The field is possibly going to be created, which is the meaning I intend to convey. I agree it is not perfect, since the document itself exists.

`update` OTOH suggests that the field already exists (in the SQL meaning)--so a function called update would not create a field, while this function does.

Both are imperfect. I renamed `update_document_payload` as I suggested, which I think does the job